core/src/main/java/com/github/jsonldjava/utils/JsonUtils.java

```diff
@@ -123,6 +123,16 @@ public static Object fromReader(Reader reader) throws IOException {
             throw new JsonParseException("document doesn't start with a valid json element : "
                     + initialToken, jp.getCurrentLocation());
         }
+        
+        JsonToken t ;
+        try { t = jp.nextToken(); }
```

Comment:
This test happens after the first JSON value in the input.  For JSON-LD is has to be an object or array but the existing code has any JSON value.  This call of nextToken gets what's next after the "}" or "]".

The exception is thrown if there is trailing junk. This was determined experimentally.  There are three cases after the closing "}" and I hope the new test cases cover all the cases.

#. All whitespace or no characters at all - legal - no token present and nextToken returns null.
#. A legal JSON token is nest, e.g. a value "string", number, "{" or even "}" nextToken returns the token. Line 134/135.
#. Illegal text, something that is not a token in JSON. e.g. plain xxxx which is a tokenising error and causes JsonParseException. Line 129/130.

In some parser/tokenizers, end of file is an explicit token (it is in javacc for example)but the Jasckson parser uses null.

The slightly different exception text messages reflect this and make it possible to tell which exception is thrown.